src/NetworkManager.cpp

```diff
@@ -0,0 +1,63 @@
+#include "NetworkManager.h"
+#include "util/config.h"
+
+#include <GLFW/glfw3.h>
+
+NetworkManager::NetworkManager() {
+    Config::config->get_value(Config::str_is_server, is_server);
+    Config::config->get_value(Config::str_server_ip, server_host);
+    Config::config->get_value(Config::str_port, port);
+
+    server = is_server ? new NetworkServer(&io_service, port) : nullptr;
+    client = new NetworkClient(&io_service);
+}
+
+NetworkManager::~NetworkManager() {
+    disconnect();
+    delete server;
+    delete client;
+}
+
+void NetworkManager::connect() {
+    // Try establishing connection in separate thread.
+    std::thread(&NetworkManager::attempt_connection, this).detach();
+}
+
+void NetworkManager::disconnect() {
+    if (!io_service.stopped())
+        io_service.stop();
+    if (service_thread) {
+        service_thread->join();
+        delete service_thread;
+    }
+}
+
+void NetworkManager::attempt_connection() {
```

Comment:
If we're keeping `NetworkManager`, I'd like to leave this in here for now.  We're defining the reconnection strategy at a higher level that the actual client, which I like.  I can see this moving into `NetworkClient` in the future though.